src/cose/mac0.rs

```diff
@@ -0,0 +1,473 @@
+                    "algorithm in protected headers did not match verifier's algorithm".into(),
+                );
+            }
+        }
+
+        let payload = match (self.inner.payload.as_ref(), detached_payload.as_ref()) {
+            (None, None) => return VerificationResult::Error(Error::NoPayload),
+            (Some(attached), None) => attached,
+            (None, Some(detached)) => detached,
+            _ => return VerificationResult::Error(Error::DoublePayload),
+        };
+
+        let tag = &self.inner.tag;
+
+        // Create the signature payload ot be used later on signing.
+        let tag_payload = mac_structure_data(
+            MacContext::CoseMac0,
+            self.inner.protected.clone(),
+            external_aad.unwrap_or_default().as_ref(),
+            payload,
+        );
+
+        let mut mac = verifier.clone();
+        mac.reset();
+        mac.update(&tag_payload);
+        match mac.verify_slice(tag) {
+            Ok(()) => VerificationResult::Success,
+            Err(e) => VerificationResult::Failure(format!("tag is not authentic: {}", e)),
+        }
+    }
+
+    /// Retrieve the CWT claims set.
+    pub fn claims_set(&self) -> Result<Option<ClaimsSet>> {
+        match self.inner.payload.as_ref() {
+            None => Ok(None),
+            Some(payload) => ClaimsSet::from_slice(payload).map_or_else(
+                |e| Err(Error::UnableToDeserializeIntoClaimsSet(e)),
+                |c| Ok(Some(c)),
+            ),
+        }
+    }
+
+    /// If we are serialized as tagged.
+    pub fn tagged(&self) -> bool {
+        self.tagged
+    }
+
+    /// Set serialization to tagged.
+    pub fn set_tagged(&mut self) {
+        self.tagged = true;
+    }
+}
+
+/// Serialize [CoseMac0] by serializing the [Value].
+/// If marked as tagged, then serialize as [Value::Tag]
+impl ser::Serialize for CoseMac0 {
+    #[inline]
+    fn serialize<S: ser::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let value = self
+            .inner
+            .clone()
+            .to_cbor_value()
+            .map_err(ser::Error::custom)?;
+        serialize::serialize(
+            value,
+            if self.tagged {
+                Some(iana::CborTag::CoseMac0 as u64)
+            } else {
+                None
+            },
+            serializer,
+        )
+    }
+}
+
+/// Deserialize [CoseMac0] by first deserializing the [Value] and then using [coset::CoseMac0::from_cbor_value].
+impl<'de> Deserialize<'de> for CoseMac0 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Deserialize the input to a CBOR Value
+        let value = Value::deserialize(deserializer)?;
+        // Convert the CBOR Value to CoseMac0
+        Ok(CoseMac0 {
+            tagged: false,
+            inner: coset::CoseMac0::from_cbor_value(value).map_err(serde::de::Error::custom)?,
+        })
```

Comment:
In `coset` especially the concept of `tagged` is only present during serialization, They have this [TaggedCborSerializable](coset-0.3.8/src/sign/mod.rs:322) trait which you use when you want to serialize Cose objects as tagged or read from tagged ones. In the memory there is nowhere kept the tagged flag, like we do.  
So you can read from a tagged serialized one but if you don't explicitly call `to_tagged_vec` it will not be tagged.  More interestingly if you don't use `from_tagged_slice` to deserialize, but use instead `coset::Value::deserialize` the `is_tag` is `false`, `as_tag` is `None` and `into_tag` is `Err`. So only using `TaggedCborSerializable` you can interact with tagged values in `coset`. But even so, the actual `CoseSign1` for ex object is deserialized ok from tagged bytes even though the Value is not seen as tagged, it knows to omit the tag part.

In our code we explicitly call `set_tagged` on `PreapredCoseSign1` obejects so it will be serialized correctly. In a test the encoded DeviceResponse contains `CoseSign1` which is NOT tagged, in that case we need to serialize it as such, so only when we explicitly desire tagged serialization we call that method, somehow similar to `coset`'s case.

Let me know if we are ok in this case.

Reply:
Unfortunately this doesn't work well. If you use serde_cbor's serializer/deserializer with ciborium tags then the tags are ignored, and if you use ciboriums serializer with serde_cbor's tags then the tags are ignored. We shouldn't be losing information just by parsing the bytes.

This is the biggest problem that we have with using both of these cbor crates in the same library, as `serde` has no native support for tagged data.

Reply:
migrated to `ciborium` and removed `serde_cbor` dep. `serde` remains because we serialize to `json` to